<!DOCTYPE html>
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=US-ASCII">
<meta name="generator" content="hevea 2.01">
<style type="text/css">
.li-itemize{margin:1ex 0ex;}
.li-enumerate{margin:1ex 0ex;}
.dd-description{margin:0ex 0ex 1ex 4ex;}
.dt-description{margin:0ex;}
.toc{list-style:none;}
.footnotetext{margin:0ex; padding:0ex;}
div.footnotetext P{margin:0px; text-indent:1em;}
.thefootnotes{text-align:left;margin:0ex;}
.dt-thefootnotes{margin:0em;}
.dd-thefootnotes{margin:0em 0em 0em 2em;}
.footnoterule{margin:1em auto 1em 0px;width:50%;}
.caption{padding-left:2ex; padding-right:2ex; margin-left:auto; margin-right:auto}
.title{margin:2ex auto;text-align:center}
.titlemain{margin:1ex 2ex 2ex 1ex;}
.titlerest{margin:0ex 2ex;}
.center{text-align:center;margin-left:auto;margin-right:auto;}
.flushleft{text-align:left;margin-left:0ex;margin-right:auto;}
.flushright{text-align:right;margin-left:auto;margin-right:0ex;}
div table{margin-left:inherit;margin-right:inherit;margin-bottom:2px;margin-top:2px}
td table{margin:auto;}
table{border-collapse:collapse;}
td{padding:0;}
.cellpadding0 tr td{padding:0;}
.cellpadding1 tr td{padding:1px;}
pre{text-align:left;margin-left:0ex;margin-right:auto;}
blockquote{margin-left:4ex;margin-right:4ex;text-align:left;}
td p{margin:0px;}
.boxed{border:1px solid black}
.textboxed{border:1px solid black}
.vbar{border:none;width:2px;background-color:black;}
.hbar{border:none;height:2px;width:100%;background-color:black;}
.hfill{border:none;height:1px;width:200%;background-color:black;}
.vdisplay{border-collapse:separate;border-spacing:2px;width:auto; empty-cells:show; border:2px solid red;}
.vdcell{white-space:nowrap;padding:0px; border:2px solid green;}
.display{border-collapse:separate;border-spacing:2px;width:auto; border:none;}
.dcell{white-space:nowrap;padding:0px; border:none;}
.dcenter{margin:0ex auto;}
.vdcenter{border:solid #FF8000 2px; margin:0ex auto;}
.minipage{text-align:left; margin-left:0em; margin-right:auto;}
.marginpar{border:solid thin black; width:20%; text-align:left;}
.marginparleft{float:left; margin-left:0ex; margin-right:1ex;}
.marginparright{float:right; margin-left:1ex; margin-right:0ex;}
.theorem{text-align:left;margin:1ex auto 1ex 0ex;}
.part{margin:2ex auto;text-align:center}
</style>
<title>StarBEAST_tutorial</title>
</head>
<body >
<!--HEVEA command line is: hevea StarBEAST_tutorial.tex -->
<!--CUT STYLE article--><!--CUT DEF section 1 --><table class="title"><tr><td style="padding:1ex"><h1 class="titlemain">*BEAST in BEAST 2.0<br>
Estimating Species Trees from Multilocus Data</h1><h3 class="titlerest">Joseph Heled, Remco Bouckaert, Alexei J Drummond and Walter Xie</h3></td></tr>
</table>
<!--TOC section id=sec1 Introduction-->
<h2 id="sec1" class="section">1&#XA0;&#XA0;Introduction</h2><!--SEC END --><p>In this tutorial we describe a full Bayesian framework for species tree estimation. The statistical methodology described in this tutorial is known by the acronym *BEAST (pronounced "star beast") [<a href="#Heled%3A2010fk">2</a>].</p><p>You will need the following software at your disposal:</p><ul class="itemize"><li class="li-itemize"><span style="font-weight:bold">BEAST</span> - this package contains the BEAST program, BEAUti, TreeAnnotator and other utility programs. This tutorial is written for BEAST v2.2.x, which has support for multiple partitions. It is available for download from <br>
<span style="font-family:monospace">http://beast2.cs.auckland.ac.nz/</span>.
</li><li class="li-itemize"><span style="font-weight:bold">Tracer</span> - this program is used to explore the output of BEAST (and other Bayesian MCMC programs). It graphically and
quantitively summarizes the distributions of continuous parameters and provides diagnostic information. At the time of
writing, the current version is v1.6. It is available for download from <span style="font-family:monospace">http://beast.bio.ed.ac.uk/</span>.
</li><li class="li-itemize"><span style="font-weight:bold">FigTree</span> - this is an application for displaying and printing molecular phylogenies, in particular those obtained using
BEAST. At the time of writing, the current version is v1.4.2. It is available for download from <span style="font-family:monospace">http://tree.bio.ed.ac.uk/</span>.
</li></ul>
<!--TOC section id=sec2 *BEAST-->
<h2 id="sec2" class="section">2&#XA0;&#XA0;*BEAST</h2><!--SEC END --><p>This tutorial will guide you through the analysis of three loci sampled from 26 individuals representing nine species of pocket gophers. This is a subset of previous published data [<a href="#belfiore2008multilocus">1</a>]. The objective of this tutorial is to estimate the species tree that is most probable given the multi-individual multi-locus sequence data. The species tree has nine taxa, whereas each gene tree has 26 taxa. *BEAST 2 will co-estimate three gene trees embedded in a shared species tree [<a href="#Heled%3A2010fk">2</a>, for details].</p><p>The first step will be to convert a NEXUS file with a DATA or CHARACTERS block into a BEAST XML input file. This is done using the program BEAUti (Bayesian Evolutionary Analysis Utility). This is a user-friendly program for setting the evolutionary model and options for the MCMC analysis. The second step is to actually run BEAST using the input file that contains the data, model and settings. The final step is to explore the output of BEAST in order to diagnose problems and to summarize the results.</p><!--TOC subsection id=sec3 BEAUti-->
<h3 id="sec3" class="subsection">BEAUti</h3><!--SEC END --><p>Run BEAUti by double clicking on its icon. </p><!--TOC subsubsection id=sec4 Set up BEAUti for *BEAST-->
<h4 id="sec4" class="subsubsection">Set up BEAUti for *BEAST</h4><!--SEC END --><p>*BEAST uses a different template from the standard. This means that to use BEAUti for *BEAST, the first thing to do is change the template. Choose the File/Templates/StarBeast item. When changing a template, BEAUti deletes all previously imported data and start with a new empty template. So, if you already loaded some data, a warning message pops up indicating that this data will be lost if you switch templates.</p><blockquote class="figure"><div class="center"><div class="center"><hr style="width:80%;height:2"></div>
<img src="figures/BEAUti_selectTemplate.png.hevea.png"><div class="caption"><table style="border-spacing:6px;border-collapse:separate;" class="cellpading0"><tr><td style="vertical-align:top;text-align:left;" >Figure 1:  Select a new template in BEAUti.</td></tr>
</table></div><div class="center"><hr style="width:80%;height:2"></div></div></blockquote><!--TOC subsubsection id=sec5 Loading the NEXUS file -->
<h4 id="sec5" class="subsubsection">Loading the NEXUS file </h4><!--SEC END --><p>*BEAST is a multi-individual, multi-locus method method. The data for each
locus is stored as one alignment in its own NEXUS file. Taxa names in each
alignment have to be unique, but duplicates across alignments are fine.
</p><p>To load a NEXUS format alignment, simply select the <span style="font-family:monospace">Import
Alignment</span> option from the File menu: </p><p>Select three files called <span style="font-family:monospace">26.nex, 29.nex, 47.nex</span> by holding <span style="font-family:monospace">shift</span> key. 
You can find the files in the <span style="font-family:monospace">examples/nexus</span> directory in the directory where BEAST was installed. 
Each file contains an alignment of sequences of from an independent locus. The <span style="font-family:monospace">26.nex</span> looks like this (content has been truncated):</p><pre class="verbatim">#NEXUS
[TBO26oLong]
BEGIN DATA;
 DIMENSIONS  NTAX =26 NCHAR=614;
 FORMAT DATATYPE = DNA GAP = - MISSING = ?;
 MATRIX 
 Orthogeomys_heterodus        ATTCTAGGCAAAAAGAGCAATGC ...
 Thomomys_bottae_awahnee_a    ????????????????????ATGCTG ...
 Thomomys_bottae_awahnee_b    ????????????????????ATGCTG ...
 Thomomys_bottae_xerophilus   ????????????????????ATGCTG ...
 Thomomys_bottae_cactophilus  ????????????????AGCAATGCT ...

         ... ...

;
END;
</pre><p><br>
</p><p>Once loaded, the three partitions are displayed in the main panel.
You can double click any alignment (partition) to show its detail.</p><blockquote class="figure"><div class="center"><div class="center"><hr style="width:80%;height:2"></div>
<img src="figures/BEAUti_DataPartitions.hevea.png"><div class="caption"><table style="border-spacing:6px;border-collapse:separate;" class="cellpading0"><tr><td style="vertical-align:top;text-align:left;" >Figure 2:  Data partition panel after loading alignments.</td></tr>
</table></div><div class="center"><hr style="width:80%;height:2"></div></div></blockquote><p>For multi-locus analyses, BEAST can link or unlink substitutions models across the loci by clicking buttons on the top of <span style="font-weight:bold">Partitions</span> panel. The default of *BEAST is unlinking all models: substitution model, clock model, and tree models. Note that you should only unlink the tree model across data partitions that are actually genetically unlinked. For example, in most organisms all the mitochondrial genes are effectively linked due to a lack of recombination and they should be set up to use the same tree model in a *BEAST analysis. </p><!--TOC subsubsection id=sec6 Import trait(s) from a mapping file to fire *BEAST-->
<h4 id="sec6" class="subsubsection">Import trait(s) from a mapping file to fire *BEAST</h4><!--SEC END --><p>Each taxon in a *BEAST analysis is associated with a species. Typically the
species name is already embedded inside the taxon. The species name should be
easy to extract; place it either at the beginning or the end, separated by a
&#X201C;special&#X201D; character which does not appear in names. For example,
<span style="font-family:monospace">aria_334259, coast_343436</span> (using an underscore) or
<span style="font-family:monospace">10x017b.wrussia, 2x305b.eastis</span> (using a dot).
</p><p>We need to tell BEAUti somehow which lineages in the alignments go with taxa in the species tree. Select the Taxon Set panel, and a list of taxa from the alignments is shown together with a default guess by BEAUti. In this case, the guess is not very good, so we want to change this. You can manually change each of the entries in the table, or press the guess button and a dialog is shown where you can choose from several ways to try to detect the taxon from the name of the lineages, or have a mapping stored in a file. In this case, splitting the name on the underscore character (&#X2019;_&#X2019;) and selecting the second group will give us the mapping that we need.</p><blockquote class="figure"><div class="center"><div class="center"><hr style="width:80%;height:2"></div>
<img src="figures/BEAUti_Guess_Taxonsets.hevea.png"><div class="caption"><table style="border-spacing:6px;border-collapse:separate;" class="cellpading0"><tr><td style="vertical-align:top;text-align:left;" >Figure 3:  Selecting taxon sets in BEAUti using the guess dialog from the taxon set panel.</td></tr>
</table></div><div class="center"><hr style="width:80%;height:2"></div></div></blockquote><p>Alternatively, the mapping can be read from a trait file.
A proper trait file is tab delimited. The first row is always <span style="font-family:monospace">traits</span> followed by the keyword <span style="font-family:monospace">species</span> in the second column and separated by tab. The rest of the rows map each individual taxon name to a species name: the taxon name in the first column and species name in the second column separated by tab. For example:</p><pre class="verbatim">traits species
taxon1 speciesA
taxon2 speciesA
taxon3 speciesB
... ...
</pre><!--TOC subsubsection id=sec7 Setting the substitution model-->
<h4 id="sec7" class="subsubsection">Setting the substitution model</h4><!--SEC END --><p>The next thing to do is to click on the <span style="font-weight:bold">Site Model</span> tab at the top of the
main window. This will reveal the evolutionary model settings for
BEAST. Exactly which options appear depend on whether the data are
nucleotides, or amino acids, or binary data, or general data.
The settings that will appear after loading the data set will
be the default values so we need to make some changes. </p><p>Most of the models should be familiar to you. For this analysis, we
will select each substitution model listed on the 
left side in turn to make the following change: select HKY for substitution model
and <span style="font-weight:bold">Empirical</span> for the <span style="font-weight:bold">Frequencies</span>. 
</p><blockquote class="figure"><div class="center"><div class="center"><hr style="width:80%;height:2"></div>
<img src="figures/BEAUti_Site_Model.hevea.png"><div class="caption"><table style="border-spacing:6px;border-collapse:separate;" class="cellpading0"><tr><td style="vertical-align:top;text-align:left;" >Figure 4:  Setting up substitution and site models for the gopher alignments.</td></tr>
</table></div><div class="center"><hr style="width:80%;height:2"></div></div></blockquote><!--TOC subsubsection id=sec8 Setting the clock model-->
<h4 id="sec8" class="subsubsection">Setting the clock model</h4><!--SEC END --><p>Second, click on the <span style="font-weight:bold">Clock Models</span> tab at the top of the
main window. In this analysis, we use the <span style="font-weight:bold">Strict Clock</span> molecular clock model as default.
Your model options should now look like this: </p><blockquote class="figure"><div class="center"><div class="center"><hr style="width:80%;height:2"></div>
<img src="figures/BEAUti_ClockModel1.png.hevea.png"><p><img src="figures/BEAUti_ClockModel2.png.hevea.png"></p><p><img src="figures/BEAUti_ClockModel3.png.hevea.png"></p><div class="caption"><table style="border-spacing:6px;border-collapse:separate;" class="cellpading0"><tr><td style="vertical-align:top;text-align:left;" >Figure 5:  Setting up clock models for the gopher alignments.</td></tr>
</table></div><div class="center"><hr style="width:80%;height:2"></div></div></blockquote><p>The <span style="font-weight:bold">Estimate</span> check box is unchecked for the first clock model and checked for the rest clock models, because we wish to estimate the substitution rate of each subsequent locus relative to the first locus whose rate is fixed to 1.0. </p><!--TOC subsubsection id=sec9 Multi Species Coalescent-->
<h4 id="sec9" class="subsubsection">Multi Species Coalescent</h4><!--SEC END --><p>The <span style="font-weight:bold">Multi Species Coalescent</span> panel allows settings to the multi species coalescent model to be specified for each tree. *BEAST has a different tree prior panel where users can only configure the species tree prior not gene tree priors (which are automatically specified by the multispecies coalescent). Currently, we have two species tree priors: <span style="font-weight:bold">Yule Process</span> and <span style="font-weight:bold">Birth-Death Process</span>; and three population size models: <span style="font-weight:bold">Piecewise linear and constant root</span>, <span style="font-weight:bold">Piecewise linear</span>, and <span style="font-weight:bold">Piecewise constant</span>. 
In this analysis, we use piecewise linear and constant root.</p><p>The <span style="font-weight:bold">Ploidy</span> item determines the type of sequence (mitochondrial, nuclear, X, Y). This matters since different modes of inheritance gives rise to different effective population sizes. In this analysis, we simply use a random starting tree. </p><blockquote class="figure"><div class="center"><div class="center"><hr style="width:80%;height:2"></div>
<img src="figures/BEAUti_MSP.hevea.png"><div class="caption"><table style="border-spacing:6px;border-collapse:separate;" class="cellpading0"><tr><td style="vertical-align:top;text-align:left;" >Figure 6:  Setting up multi species coalescent parameters.</td></tr>
</table></div><div class="center"><hr style="width:80%;height:2"></div></div></blockquote><!--TOC subsubsection id=sec10 Priors and Operators-->
<h4 id="sec10" class="subsubsection">Priors and Operators</h4><!--SEC END --><p>The <span style="font-weight:bold">Priors</span> panel allows priors to be specified for each parameter in the model. 
The <span style="font-weight:bold">Operators</span> panel (hidden) is used to configure technical settings that affect the efficiency of the MCMC program. 
We leave these two panels unchanged in this analysis.</p><!--TOC subsubsection id=sec11 Setting the MCMC options -->
<h4 id="sec11" class="subsubsection">Setting the MCMC options </h4><!--SEC END --><p>The next tab, <span style="font-weight:bold">MCMC</span>, provides more general
settings to control the length of the MCMC and the file names. </p><p>Firstly we have the <span style="font-weight:bold">Length of chain</span>. This is the number of
steps the MCMC will make in the chain before finishing. The appropriate length of the chain depends on the size of the data set, the complexity of the
model and the accuracy of the answer required. The default value of 10,000,000
is entirely arbitrary and should be adjusted according to the size
of your data set. For this data set let&#X2019;s set the <span style="font-family:monospace">chain
length</span> to <U>5,000,000</U> as this will run reasonably quickly on most modern
computers (a few minutes).</p><p>The next options specify how often the parameter values in the Markov
chain should be displayed on the screen and recorded in the log file.
The screen output is simply for monitoring the programs progress so
can be set to any value (although if set too small, the sheer quantity
of information being displayed on the screen will actually slow the
program down). For the log file, the value should be set relative
to the total length of the chain. Sampling too often will result in
very large files with little extra benefit in terms of the precision
of the analysis. Sample too infrequently and the log file will not
contain much information about the distributions of the parameters. 
You probably want to aim to store no more than 10,000 samples so this should be
set to no less than chain length / 10,000.</p><p>For this exercise we will set the <span style="font-family:monospace">screen log</span> to <U>10,000</U> and the <span style="font-family:monospace">trace log</span> to <U>5,000</U>. The final two
options give the file names of the log files for the sampled parameters and
the trees. These will be set to a default based on the name of the
imported NEXUS file. </p><blockquote class="figure"><div class="center"><div class="center"><hr style="width:80%;height:2"></div>
<img src="figures/BEAUti_MCMC.hevea.png"><div class="caption"><table style="border-spacing:6px;border-collapse:separate;" class="cellpading0"><tr><td style="vertical-align:top;text-align:left;" >Figure 7:  Setting up the MCMC paremeters.</td></tr>
</table></div><div class="center"><hr style="width:80%;height:2"></div></div></blockquote><p>If you are using windows then we suggest you add the suffix <span style="font-family:monospace">.txt</span> to both of these (so,
<span style="font-family:monospace">gopher.log.txt</span> and <span style="font-family:monospace">gopher.trees.txt</span>) so that Windows recognizes
these as text files. </p><!--TOC subsubsection id=sec12 Generating the BEAST XML file -->
<h4 id="sec12" class="subsubsection">Generating the BEAST XML file </h4><!--SEC END --><p>We are now ready to create the BEAST XML file. To do this, either select the <span style="font-weight:bold">File/Save</span> or <span style="font-weight:bold">File/Save As</span> option from the <span style="font-weight:bold">File</span> menu. Check the default priors setting and click <span style="font-weight:bold">Continue</span>. Save the file with an appropriate name (we usually end the filename with <span style="font-family:monospace">.xml</span>, i.e., <span style="font-family:monospace">gopher.xml</span>). We are now ready to run the file through BEAST. </p><!--TOC subsection id=sec13 Running BEAST -->
<h3 id="sec13" class="subsection">Running BEAST </h3><!--SEC END --><p>Now run BEAST and when it asks for an input file, provide your newly
created XML file as input by click <span style="font-weight:bold">Choose File ...</span>, and then click <span style="font-weight:bold">Run</span>. </p><blockquote class="figure"><div class="center"><div class="center"><hr style="width:80%;height:2"></div>
<img src="figures/BEAST.hevea.png"><div class="caption"><table style="border-spacing:6px;border-collapse:separate;" class="cellpading0"><tr><td style="vertical-align:top;text-align:left;" >Figure 8:  Launching BEAST.</td></tr>
</table></div><div class="center"><hr style="width:80%;height:2"></div></div></blockquote><p>BEAST will then run until it has finished
reporting information to the screen. The actual results files are
saved to the disk in the same location as your input file. The output to the screen will
look something like this: </p><pre class="verbatim"><span style="font-size:xx-small">                        BEAST v2.2.0, 2002-2014
             Bayesian Evolutionary Analysis Sampling Trees
                       Designed and developed by
 Remco Bouckaert, Alexei J. Drummond, Andrew Rambaut &amp; Marc A. Suchard
                                    
                     Department of Computer Science
                         University of Auckland
                        remco@cs.auckland.ac.nz
                        alexei@cs.auckland.ac.nz
                                    
                   Institute of Evolutionary Biology
                        University of Edinburgh
                           a.rambaut@ed.ac.uk
                                    
                    David Geffen School of Medicine
                 University of California, Los Angeles
                           msuchard@ucla.edu
                                    
                      Downloads, Help &amp; Resources:
                           http://beast2.org/
                                    
  Source code distributed under the GNU Lesser General Public License:
                   http://github.com/CompEvol/beast2
                                    
                           BEAST developers:
   Alex Alekseyenko, Trevor Bedford, Erik Bloomquist, Joseph Heled, 
 Sebastian Hoehna, Denise Kuehnert, Philippe Lemey, Wai Lok Sibon Li, 
Gerton Lunter, Sidney Markowitz, Vladimir Minin, Michael Defoin Platel, 
                 Oliver Pybus, Chieh-Hsi Wu, Walter Xie
                                    
                               Thanks to:
          Roald Forsberg\includeimage, Beth Shapiro and Korbinian Strimmer


... ...


        4980000     -3827.2093       279.2        -4293.2486        22.1146 24s/Msamples
        4990000     -3842.9689       276.3        -4295.6213        17.6877 24s/Msamples
        5000000     -3798.1962       278.2        -4280.7865        22.0904 24s/Msamples

Operator                                           Tuning    #accept    #reject      Pr(m)  Pr(acc|m)
NodeReheight(Reheight.t:Species)                        -     205317     486572     0.1386     0.2967 
ScaleOperator(popSizeBottomScaler.t:Species)       0.1902      10224      26338     0.0074     0.2796 
ScaleOperator(popMeanScale.t:Species)              0.4890       5795      16242     0.0044     0.2630 
UpDownOperator(updown.all.Species)                 0.5029      38885     108717     0.0295     0.2634 
ScaleOperator(YuleBirthRateScaler.t:Species)       0.2416       7155      15125     0.0044     0.3211 
ScaleOperator(treeScaler.t:26)                     0.7580       8510      42962     0.0102     0.1653 
ScaleOperator(treeRootScaler.t:26)                 0.4424      11487      39392     0.0102     0.2258 
Uniform(UniformOperator.t:26)                           -     287517     225590     0.1024     0.5603 
SubtreeSlide(SubtreeSlide.t:26)                    0.5560        418     256419     0.0512     0.0016 Try decreasing size to about 0.278
Exchange(narrow.t:26)                                   -     115438     140923     0.0512     0.4503 
Exchange(wide.t:26)                                     -       1363      49988     0.0102     0.0265 
WilsonBalding(WilsonBalding.t:26)                       -       1882      49890     0.0102     0.0364 
ScaleOperator(StrictClockRateScaler.c:29)          0.4744      13355      37595     0.0102     0.2621 
ScaleOperator(treeScaler.t:29)                     0.7567       8137      42785     0.0102     0.1598 
ScaleOperator(treeRootScaler.t:29)                 0.4101      12191      38887     0.0102     0.2387 
Uniform(UniformOperator.t:29)                           -     296262     216280     0.1024     0.5780 
SubtreeSlide(SubtreeSlide.t:29)                    0.5724        339     255443     0.0512     0.0013 Try decreasing size to about 0.286
Exchange(narrow.t:29)                                   -     117912     138109     0.0512     0.4606 
Exchange(wide.t:29)                                     -       2229      49303     0.0102     0.0433 
WilsonBalding(WilsonBalding.t:29)                       -       2459      49001     0.0102     0.0478 
UpDownOperator(updown.29)                          0.7793      10314      40694     0.0102     0.2022 
ScaleOperator(StrictClockRateScaler.c:47)          0.5490      12974      38097     0.0102     0.2540 
ScaleOperator(treeScaler.t:47)                     0.7363       8746      42523     0.0102     0.1706 
ScaleOperator(treeRootScaler.t:47)                 0.4742       7774      43670     0.0102     0.1511 
Uniform(UniformOperator.t:47)                           -     276942     234361     0.1024     0.5416 
SubtreeSlide(SubtreeSlide.t:47)                    0.5044        307     255119     0.0512     0.0012 Try decreasing size to about 0.252
Exchange(narrow.t:47)                                   -      91846     164747     0.0512     0.3579 
Exchange(wide.t:47)                                     -        731      50281     0.0102     0.0143 
WilsonBalding(WilsonBalding.t:47)                       -        937      50377     0.0102     0.0183 
UpDownOperator(updown.47)                          0.7397      10853      40537     0.0102     0.2112 
UpDownOperator(strictClockUpDownOperator.c:47)     0.7278       9987      41373     0.0102     0.1945 
UpDownOperator(strictClockUpDownOperator.c:29)     0.7761      10040      40724     0.0102     0.1978 
ScaleOperator(KappaScaler.s:26)                    0.3401        469       1197     0.0003     0.2815 
ScaleOperator(KappaScaler.s:29)                    0.2955        436       1231     0.0003     0.2615 
ScaleOperator(KappaScaler.s:47)                    0.3199        379       1298     0.0003     0.2260 
ScaleOperator(popSizeTopScaler.t:Species)          0.1910      23665      54936     0.0157     0.3011 

     Tuning: The value of the operator's tuning parameter, or '-' if the operator can't be optimized.
    #accept: The total number of times a proposal by this operator has been accepted.
    #reject: The total number of times a proposal by this operator has been rejected.
      Pr(m): The probability this operator is chosen in a step of the MCMC (i.e. the normalized weight).
  Pr(acc|m): The acceptance probability (#accept as a fraction of the total proposals for this operator).


Total calculation time: 135.584 seconds
End likelihood: -3798.196207797654
</span></pre><!--TOC subsection id=sec14 Analyzing the results-->
<h3 id="sec14" class="subsection">Analyzing the results</h3><!--SEC END --><p>Run the program called <span style="font-weight:bold">Tracer</span> to analyze the output of BEAST. When the main
window has opened, choose <span style="font-weight:bold">Import Trace File...</span> from the <span style="font-weight:bold">File</span> menu and select the file that
BEAST has created called <span style="font-family:monospace">gopher.log</span>.
You should now see a window like in Figure <a href="#fig.tracer1">9</a>.</p><blockquote class="figure"><div class="center"><div class="center"><hr style="width:80%;height:2"></div>
<img src="figures/Tracer1.hevea.png"><div class="caption"><table style="border-spacing:6px;border-collapse:separate;" class="cellpading0"><tr><td style="vertical-align:top;text-align:left;" >Figure 9:  Tracer with the gopher data.</td></tr>
</table></div><div class="center"><hr style="width:80%;height:2"></div></div></blockquote><p>Remember that MCMC is a stochastic algorithm so the actual numbers will not be exactly the same.</p><p>On the left hand side is a list of the different quantities that BEAST has logged. There are traces for the posterior (this
is the log of the product of the tree likelihood and the prior probabilities), and the continuous parameters. Selecting a trace
on the left brings up analyses for this trace on the right hand side depending on tab that is selected. When first opened, the
&#X2018;posterior&#X2019; trace is selected and various statistics of this trace are shown under the Estimates tab.
In the top right of the window is a table of calculated statistics for the selected trace. </p><p>Tracer will plot a (marginal posterior) distribution for the selected parameter and also give you statistics such as the mean and median. The <span style="font-family:monospace">95% HPD lower</span> or <span style="font-family:monospace">upper</span> stands for <span style="font-style:italic">highest posterior density interval</span> and represents the most compact interval on the selected parameter that contains 95% of the posterior probability. It can be thought of as a Bayesian analog to a confidence interval. </p><p>Select the <span style="font-family:monospace">treeModel.rootHeight</span> parameter and the next three (hold shift whilst selecting). This will show a display of the
age of the root and the three gene trees. If you switch the tab at the top of the window to <span style="font-weight:bold">Marginal Density</span> then you will get a plot of the marginal posterior densities of each of these date estimates overlayed,
as shown in Figure <a href="#fig.tracer2">10</a>.</p><blockquote class="figure"><div class="center"><div class="center"><hr style="width:80%;height:2"></div>
<img src="figures/Tracer2.hevea.png"><div class="caption"><table style="border-spacing:6px;border-collapse:separate;" class="cellpading0"><tr><td style="vertical-align:top;text-align:left;" >Figure 10:  Tracer showing the root heights of the lineage trees.</td></tr>
</table></div><div class="center"><hr style="width:80%;height:2"></div></div></blockquote><!--TOC subsection id=sec15 Obtaining an estimate of the phylogenetic tree-->
<h3 id="sec15" class="subsection">Obtaining an estimate of the phylogenetic tree</h3><!--SEC END --><p>BEAST also produces a sample of plausible trees. 
These can be summarized using the program <span style="font-weight:bold">TreeAnnotator</span>. This will take the set of trees and identify a single tree that best represents the posterior distribution. It will then annotate this selected tree topology with the mean ages of all the
nodes as well as the 95% HPD interval of divergence times for each clade in the selected tree. It will also calculate the posterior clade probability for each
node. Run the <span style="font-weight:bold">TreeAnnotator</span> program and set it up to look like in Figure <a href="#fig.TreeAnnotator">11</a>.</p><blockquote class="figure"><div class="center"><div class="center"><hr style="width:80%;height:2"></div>
<img src="figures/TreeAnnotator.hevea.png"><div class="caption"><table style="border-spacing:6px;border-collapse:separate;" class="cellpading0"><tr><td style="vertical-align:top;text-align:left;" >Figure 11:  Using TreeAnnotator to summarise the tree set.</td></tr>
</table></div><div class="center"><hr style="width:80%;height:2"></div></div></blockquote><p>The burnin is the number of trees to remove from the start of the sample. Unlike <span style="font-weight:bold">Tracer</span> which specifies the number of steps as a burnin, in <span style="font-weight:bold">TreeAnnotator</span> you need to specify the actual number of trees. For this run, we use the default setting.</p><p>The <span style="font-weight:bold">Posterior probability limit</span> option specifies a limit such that if a node is found at less than this frequency in the sample of trees (i.e., has a posterior probability less than this limit), it will not be annotated. The default of 0.5 means that only nodes seen in the majority of trees will be annotated. Set this to zero to annotate all nodes.</p><p>For <span style="font-weight:bold">Target tree type</span> you can either choose a specific tree from a file or ask TreeAnnotator to find a tree in your sample. The default option, <span style="font-weight:bold">Maximum clade credibility tree</span>, finds the tree with the highest product of the posterior probability of all its nodes.</p><p>Choose <span style="font-weight:bold">Mean heights</span> for node heights. This sets the heights (ages) of each node in the tree to the mean height across the entire sample of trees for that clade.</p><p>For the input file, select the trees file that BEAST created (by default this will be called <span style="font-family:monospace">gopher.species.trees</span>) and select a file for the output (here we called it <span style="font-family:monospace">gopher.species.tree</span>).</p><p>Now press <span style="font-family:monospace">Run</span> and wait for the program to finish.</p><!--TOC subsection id=sec16 Viewing the Sprecies Tree-->
<h3 id="sec16" class="subsection">Viewing the Sprecies Tree</h3><!--SEC END --><p>Finally, we can look at the tree in another program called <span style="font-weight:bold">FigTree</span>. Run this program, and open
the <span style="font-family:monospace">gopher.species.tree</span> file by using the Open command in the File menu. The tree should appear.
You can now try selecting some of the options in the control panel on the left. Try selecting
<span style="font-weight:bold">Node Bars</span> to get node age error bars. Also turn on <span style="font-weight:bold">Branch Labels</span> and select <span style="font-weight:bold">posterior</span> to get
it to display the posterior probability for each node. Under <span style="font-weight:bold">Appearance</span> you can also tell FigTree
to colour the branches by the rate.
You should end up with something like Figure <a href="#fig.figtree">12</a>.</p><blockquote class="figure"><div class="center"><div class="center"><hr style="width:80%;height:2"></div>
<img src="figures/figtree.pdf.hevea.png"><div class="caption"><table style="border-spacing:6px;border-collapse:separate;" class="cellpading0"><tr><td style="vertical-align:top;text-align:left;" >Figure 12:  Figtree representation of the species tree.</td></tr>
</table></div><div class="center"><hr style="width:80%;height:2"></div></div></blockquote><p>Alternatively, you can load the species tree set into DensiTree and set it up as follows.</p><ul class="itemize"><li class="li-itemize">
Set burn-in to 500. The tree should not be collapsed any more.
</li><li class="li-itemize">Show a root-canal tree to guide the eye. Perhaps, the intensity of the trees is not large enough, so you might want to increase the intensity by clicking the icon in the button bar.
</li><li class="li-itemize">Show clades, their mean and 95% HPD graphically, and posterior support using text. Now, too many clades are shown, and most are not of interest. Select &#X2019;Selected only&#X2019;, then open the clade toolbar (menu Window/View clade toolbar), and select only highly supported clades. Also, select the clade consisting of heterodus, bottea, umbinus and townsendii, to show that heterodus is an outgroup,
but there is some support (over 16%) that it is not.
</li><li class="li-itemize">Drag the clade monticola and idahoensis up so that the 95% HPD bar does not overlap with the one for mazama, monticola, idahoensis and talpoidis. Increase font size of the label for better readability.
</li></ul><p>The image should look something like Figure <a href="#fig.DensiTree">13</a></p><blockquote class="figure"><div class="center"><div class="center"><hr style="width:80%;height:2"></div>
<img src="figures/DensiTree.png.hevea.png">
<div class="caption"><table style="border-spacing:6px;border-collapse:separate;" class="cellpading0"><tr><td style="vertical-align:top;text-align:left;" >Figure 13:  DensiTree representation of the species tree.</td></tr>
</table></div>
<div class="center"><hr style="width:80%;height:2"></div></div></blockquote><p>Exercise: There is about 75% support for heterodus to be an outgroup, and about 17% for heterodus to be in a clade with bottea, umbinus and townsendii. Can you explain where the other 8% went?</p><p>DensiTree can be used to show the branch widths of summary 
tree from tree annotator as population sizes. 
Under &#X2018;Line Width&#X2019; in DensiTree, choose &#X2018;BY_METADATA_NUMBER&#X2019; for the bottom and for the top,
and choose numbers 2 and 3 in the &#X2019;top&#X2019; and &#X2019;bottom&#X2019; spinner.
Left, the bottom represents dmv1, the top dmv2 in the summary tree, which do not quite match 
in areas with little posterior support for the clades (see Figure <a href="#fig.DensiTree">13</a> to
see which clades have little support).</p><p>Right, the top is matched up with the bottom of the branch above, using the &#X2018;Make fit to bottom&#X2019;
option for the top. This looks a bit prettier, but may not be quite accurate.</p><blockquote class="figure"><div class="center"><div class="center"><hr style="width:80%;height:2"></div>
<img src="figures/species_population1.pdf.hevea.png">
<div class="center"><hr style="width:80%;height:2"></div></div></blockquote><blockquote class="figure"><div class="center"><div class="center"><hr style="width:80%;height:2"></div>
<img src="figures/species_population2.pdf.hevea.png">
<div class="center"><hr style="width:80%;height:2"></div></div></blockquote><p>Alternatively, a consensus tree can be generated by biopy (<span style="font-family:monospace">http://code.google.com/p/biopy/</span>)
with using 1-norm left, and 2-norm right.</p><blockquote class="figure"><div class="center"><div class="center"><hr style="width:80%;height:2"></div>
<img src="figures/species_population4.pdf.hevea.png">
<div class="center"><hr style="width:80%;height:2"></div></div></blockquote><blockquote class="figure"><div class="center"><div class="center"><hr style="width:80%;height:2"></div>
<img src="figures/species_population5.pdf.hevea.png">
<div class="center"><hr style="width:80%;height:2"></div></div></blockquote><p>Showing all consensus trees with population widths
(use By Metadata Pattern, for top and bottom and use </p><pre class="verbatim"> .*dmv=.([^,]*).*</pre><p> 
for the bottom pattern and </p><pre class="verbatim"> .*dmv=.[^,]*,([^\}]*).*</pre><p> for the top pattern.
This gives us this visualisation:</p><blockquote class="figure"><div class="center"><div class="center"><hr style="width:80%;height:2"></div>
<img src="figures/species_population3.pdf.hevea.png">
<div class="center"><hr style="width:80%;height:2"></div></div></blockquote><!--TOC subsection id=sec17 Comparing your results to the prior-->
<h3 id="sec17" class="subsection">Comparing your results to the prior</h3><!--SEC END --><p>Using BEAUti, set up the same analysis but under the MCMC options, select the <span style="font-weight:bold">Sample from prior only</span> option. This will allow you to visualize the full prior distribution in the absence of your sequence data. Summarize the trees from the full prior
distribution and compare the summary to the posterior summary tree.</p><!--TOC section id=sec18 References-->
<h2 id="sec18" class="section">References</h2><!--SEC END --><dl class="thebibliography"><dt class="dt-thebibliography">
<a id="belfiore2008multilocus">[1]</a></dt><dd class="dd-thebibliography">
N.M. Belfiore, L.&#XA0;Liu, and C.&#XA0;Moritz.
Multilocus phylogenetics of a rapid radiation in the genus Thomomys
(Rodentia: Geomyidae).
<em>Systematic Biology</em>, 57(2):294, 2008.</dd><dt class="dt-thebibliography"><a id="Heled:2010fk">[2]</a></dt><dd class="dd-thebibliography">
Joseph Heled and Alexei&#XA0;J Drummond.
Bayesian inference of species trees from multilocus data.
<em>Mol Biol Evol</em>, 27(3):570&#X2013;80, Mar 2010.</dd></dl><!--CUT END -->
<!--HTMLFOOT-->
<!--ENDHTML-->
<!--FOOTER-->
<hr style="height:2"><blockquote class="quote"><em>This document was translated from L<sup>A</sup>T<sub>E</sub>X by
</em><a href="http://hevea.inria.fr/index.html"><em>H</em><em><span style="font-size:small"><sup>E</sup></span></em><em>V</em><em><span style="font-size:small"><sup>E</sup></span></em><em>A</em></a><em>.</em></blockquote></body>
</html>
